app/ts/LiquidityProvidingUI/Liquidity.tsx

```diff
@@ -215,47 +215,57 @@ export const Liquidity = ({ maybeReadClient, maybeWriteClient, universe, reputat
+		if (selectedMarket.deepValue === undefined) return
+		marketData.deepValue = await fetchMarketData(maybeReadClient.deepValue, selectedMarket.deepValue)
+		const disputeWindowAddress = await getDisputeWindow(maybeReadClient.deepValue, selectedMarket.deepValue)
 		if (EthereumAddress.parse(disputeWindowAddress) !== 0n) {
-			disputeWindowInfo.deepValue = await getDisputeWindowInfo(readClient, disputeWindowAddress)
+			disputeWindowInfo.deepValue = await getDisputeWindowInfo(maybeReadClient.deepValue, disputeWindowAddress)
 		}
-		forkValues.deepValue = await getForkValues(readClient, reputationTokenAddress.deepValue)
 
-		isConstantProductMarketDeployed.deepValue = await isThereAugurConstantProductmarket(readClient, marketAddress.deepValue)
+		isConstantProductMarketDeployed.deepValue = await isThereAugurConstantProductMarket(maybeReadClient.deepValue, selectedMarket.deepValue)
 
-		const writeClient = maybeWriteClient.deepPeek()
-		if (writeClient === undefined) throw new Error('missing writeClient')
-		daiApprovedForRouter.deepValue = await getAllowanceErc20Token(writeClient, DAI_TOKEN_ADDRESS, writeClient.account.address, getAugurConstantProductMarketRouterAddress())
+		if (maybeWriteClient.deepValue === undefined) return
+		daiApprovedForRouter.deepValue = await getAllowanceErc20Token(maybeWriteClient.deepValue, DAI_TOKEN_ADDRESS, maybeWriteClient.deepValue.account.address, getAugurConstantProductMarketRouterAddress())
 		const router = getAugurConstantProductMarketRouterAddress()
-		sharesApprovedToRouter.deepValue = await isErc1155ApprovedForAll(readClient, AUGUR_SHARE_TOKEN, writeClient.account.address, router)
+		sharesApprovedToRouter.deepValue = await isErc1155ApprovedForAll(maybeReadClient.deepValue, AUGUR_SHARE_TOKEN, maybeWriteClient.deepValue.account.address, router)
+
+		liquidityTokens.deepValue = await getUserLpTokenIdsAndBalancesForMarket(maybeReadClient.deepValue, marketData.deepValue.marketAddress, maybeWriteClient.deepValue.account.address)
+	}
 
+	const updateShareBalances = async () => {
 		if (maybeWriteClient.deepValue === undefined) return
+		if (marketData.deepValue === undefined) return
+		if (isConstantProductMarketDeployed.deepValue !== true) return
 		const shareBalances = await getShareBalances(maybeWriteClient.deepValue, marketData.deepValue.marketAddress, maybeWriteClient.deepValue.account.address)
 		invalidBalance.deepValue = shareBalances[0]
 		noBalance.deepValue = shareBalances[1]
 		yesBalance.deepValue = shareBalances[2]
-
-		liquidityTokens.deepValue = await getUserLpTokenIdsAndBalancesForMarket(readClient, marketData.deepValue.marketAddress, maybeWriteClient.deepValue.account.address)
 	}
 
+	useSignalEffect(() => {
+		if (maybeWriteClient.deepValue === undefined) return
+		if (marketData.deepValue === undefined) return
+		if (isConstantProductMarketDeployed.deepValue !== true) return
+		updateShareBalances()
+	})
```

Comment:
See comment below on the problem with this pattern and suggestions on how to deal with it.